common/src/main/java/tachyon/security/PlainSaslServer.java

```diff
@@ -144,4 +152,47 @@ private void throwIfNotComplete() {
       throw new IllegalStateException("PLAIN authentication not completed");
     }
   }
+
+  /**
+   * PlainServerCallbackHandler is used by the SASL mechanisms to get further information
+   * to complete the authentication. For example, a SASL mechanism might uses this callback handler
+   * to do verify operation.
+   */
+  public static final class PlainServerCallbackHandler implements CallbackHandler {
+    private final AuthTypes mAuthType;
+    private final TachyonConf mConf;
+
+    public PlainServerCallbackHandler(AuthTypes authType, TachyonConf conf) {
+      mAuthType = authType;
+      mConf = conf;
+    }
+
+    @Override
+    public void handle(Callback[] callbacks) throws IOException, UnsupportedCallbackException {
+      String username = null;
+      String password = null;
+      AuthorizeCallback ac = null;
+
+      for (Callback callback : callbacks) {
+        if (callback instanceof NameCallback) {
+          NameCallback nc = (NameCallback) callback;
+          username = nc.getName();
+        } else if (callback instanceof PasswordCallback) {
+          PasswordCallback pc = (PasswordCallback) callback;
+          password = new String(pc.getPassword());
+        } else if (callback instanceof AuthorizeCallback) {
+          ac = (AuthorizeCallback) callback;
+        } else {
+          throw new UnsupportedCallbackException(callback, "Unsupport callback");
+        }
+      }
+
+      AuthenticationProvider provider =
```

Comment:
Here, a provider has to be created for every connection, and for some providers, e.g. CUSTOM, it requires expensive reflection. It seems we only need to create one during construction of `PlainServerCallbackHandler`? Because it takes only `mAuthType` and `mConf` and `#authenticate` is completely stateless?